model_signing/manifest/in_toto.py

```diff
@@ -0,0 +1,104 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This package provides functionality to convert file based manifests
+to in-toto statements. It is necessary because sigstore does not
+support arbitrary payloads in DSSE envelopes.
+"""
+import pathlib
+
+from in_toto_attestation.v1 import statement
+from in_toto_attestation.v1 import resource_descriptor
+
+from model_signing.hashing.hashing import Digest
+from model_signing.manifest import manifest
+from model_signing.manifest.manifest import FileLevelManifest
+from model_signing.manifest.manifest import FileManifestItem
+
+
+_PREDICATE_TYPE = 'model_signing/v1/manifest'
+_FILE_LEVEL_MANIFEST = 'FileLevelManifest'
+
+
+def _file_level_manifest_to_statement(
+        manifest: FileLevelManifest,
+        ) -> statement.Statement:
+    """
+    Converts a model signing FileLevelManifest to an
+    in-toto statement.
+
+    Args:
+        manifest (FileLevelManifest): the manifest to convert
+        algorithm (str): the used hash algorithm
+
+    Returns:
+        statement.Statement: the in-toto statement representing the manifest
+    """
+    subjects: list[resource_descriptor.ResourceDescriptor] = []
+    for item in manifest.items:
+        s = resource_descriptor.ResourceDescriptor(
+            name=str(item.path),
+            digest={item.digest.algorithm: item.digest.digest_hex},
+        ).pb
+        subjects.append(s)
+    return statement.Statement(
+        subjects=subjects,
+        predicate_type=_PREDICATE_TYPE,
+        predicate={'manifest_type': _FILE_LEVEL_MANIFEST},
+        )
+
+
+def _statement_to_file_level_manifest(
+        statement: statement.Statement,
+        ) -> FileLevelManifest:
+    """
+    Converts an in-toto statement to a FileLevelManifest.
+
+    Args:
+        statement (statement.Statement): the in-toto statement
+        algorithm (str): the hash algorithm used
+
+    Returns:
+        FileLevelManifest: the resutling FileLevelManifest
+    """
+    items: list[FileManifestItem] = []
+    for s in statement.pb.subject:
+        # no support for multiple hashes
+        alg, dig = list(s.digest.items())[0]
+        items.append(
+            FileManifestItem(
+                path=pathlib.Path(s.name),
+                digest=Digest(
+                    algorithm=alg,
+                    digest_value=bytes.fromhex(dig),
+                )
+            )
+        )
+    return FileLevelManifest(items)
+
+
+def manifest_to_statement(
+        model_manifest: manifest.Manifest
+        ) -> statement.Statement:
+    if isinstance(model_manifest, manifest.FileLevelManifest):
```

Comment:
Yes, we can do that in the future (before v1). 

However, before implementing any other conversion function I want to be sure that we are aligned on using in-toto in a dsse envelope wrapped into a sigstore bundle as on disk format.

Reply:
That's the problem, we need to support other on disk formats too before the v1 release.

Reply:
We should be opinionated about this and define the disk format.
The goal is to deliver something that can be easily verified by a broad user base and doesn't need to find the right verification format.
Splitting away from in-toto as disk format makes it more difficult for sigstore based signatures since they don't support arbitrary payloads in dsse envelopes.
Perhaps we also need to decide on shipping the manifest (perhaps as json) and a sigstore bundle with the model so we can provide the signature separately from the manifest making verification a bit more complicated. The advantage here is we can use the manifest directly as disk format and only need to support signing over the manifest itself. For sigstore this would still allow us to use the hashed rekord variant instead of the dsse envelope. 

In any case vaguely saying "we need to support other disk formats too before v1 release" does not really help since it is a bit like saying we need to support all formats.